### PR TITLE
Bundle resource path parameters of external function into anydata array

### LINKED ISSUE #39781: [Improvement]: Bundle all path params in the resource function to BArray in the native function
### Description

Support following Java interop binding,

The Ballerina function
```
resource function get resourceName/[BT1 p1]/[BT2 p2]/.../[BTn pn](BT1 t1, …, BTn tn) returns BRT = @java:Method {
	[name: “methodName”,]
class: “x.y.z.ClassName”,
} external;
```
The Java function
```
class: 	x.y.z.ClassName;
method: JRT methodName(Environment env, BObject client, BArray pathParams, JT1 t1, JT2 t2, …, JTn tn);
```

Here all the path param values in the Ballerina resource function are bundled to the BArray pathParams variable. BArray with anydata types.

### Describe your problem(s)

In the current support, we need input parameters for each path param defined in the Ballerina resource function. but we have a generic native function that needs to bind to any resource function and we don't know how many path params are in the resource function.

### Describe your solution(s)

Described the in the description.

### Related area

-> Runtime

### Related issue(s) (optional)

https://github.com/ballerina-platform/ballerina-lang/issues/39737

### Suggested label(s) (optional)

jballerina

### Suggested assignee(s) (optional)

@warunalakshitha

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #39781 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 1]
```diff
@@ -606,8 +610,41 @@ private void genJICallTerm(JIMethodCall callIns, int localVarOffset) {
...
+            for (BIROperand arg : pathArgs) {
+                mv.visitVarInsn(ALOAD, pathVarArrayIndex);
+                mv.visitLdcInsn((long) i);
+                mv.visitInsn(L2I);
+                this.loadVar(arg.variableDcl);
+                jvmCastGen.generateCheckCastToAnyData(mv, arg.variableDcl.type);
+                mv.visitInsn(AASTORE);
+                i++;
+            }
+
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:636]: less remove unwanted new lines

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 2]
```diff
@@ -606,8 +610,41 @@ private void genJICallTerm(JIMethodCall callIns, int localVarOffset) {
             this.mv.visitVarInsn(ALOAD, localVarOffset); // load the strand
             // load the current Module
             mv.visitFieldInsn(GETSTATIC, this.moduleInitClass, CURRENT_MODULE_VAR_NAME, GET_MODULE);
-            mv.visitMethodInsn(INVOKESPECIAL, BAL_ENV, JVM_INIT_METHOD,
-                               INIT_BAL_ENV, false);
+            mv.visitMethodInsn(INVOKESPECIAL, BAL_ENV, JVM_INIT_METHOD, INIT_BAL_ENV, false);
+        }
+
+        if (callIns.resourcePathArgs != null && !callIns.resourcePathArgs.isEmpty()) {
+            List<BIROperand> pathArgs = callIns.resourcePathArgs;
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:617]: Shall we move handle path paramm to separate function

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 3]
```diff
@@ -606,8 +610,41 @@ private void genJICallTerm(JIMethodCall callIns, int localVarOffset) {
...
+                mv.visitInsn(AASTORE);
+                i++;
+            }
+
+            mv.visitTypeInsn(NEW, ARRAY_VALUE_IMPL);
+            mv.visitInsn(DUP);
+            mv.visitVarInsn(ALOAD, pathVarArrayIndex);
+            mv.visitTypeInsn(NEW, ARRAY_TYPE_IMPL);
+            mv.visitInsn(DUP);
+            mv.visitFieldInsn(GETSTATIC, PREDEFINED_TYPES, "TYPE_ANYDATA", LOAD_ANYDATA_TYPE);
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:642]: We can use some constants

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java [hunk 4]
```diff
@@ -333,6 +334,7 @@ public class JvmSignatures {
     public static final String LINKED_HASH_SET_OP = "(L" + LINKED_HASH_SET + ";)V";
     public static final String LOAD_ANY_TYPE = "L" + ANY_TYPE + ";";
     public static final String LOAD_ANYDATA_TYPE = "L" + ANYDATA_TYPE + ";";
+    public static final String ANYDATA_ARRAY_INIT = "([L" + OBJECT + ";L" + ARRAY_TYPE + ";)V";
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java:337]:
```suggestion
    public static final String INIT_ANYDATA_ARRAY = "([L" + OBJECT + ";L" + ARRAY_TYPE + ";)V";
```

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java:337]: move this constant to other init constants

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 5]
```diff
@@ -554,8 +560,7 @@ private void genJICallTerm(JIMethodCall callIns, int localVarOffset) {
         genHandlingBlockedOnExternal(localVarOffset, blockedOnExternLabel);
         if (callIns.lhsOp != null) {
             this.mv.visitVarInsn(ALOAD, localVarOffset);
-            this.mv.visitFieldInsn(GETFIELD, STRAND_CLASS, "returnValue",
-                                   "Ljava/lang/Object;");
+            this.mv.visitFieldInsn(GETFIELD, STRAND_CLASS, "returnValue", "Ljava/lang/Object;");
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:563]: Ljava/lang/Object; Can be replace by already having constants, signatures

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java (hunk 1 above)
Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:636]: in other places too

File: tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/variable_number_of_resource_paths.bal [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods",
+        name: "getResource"
+    } external;
+
+    isolated resource function get abc/[int id]/[string path]/[string ...paths] () returns int = @java:Method {
+        'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods",
+        name: "getResource"
+    } external;
+
+    isolated resource function get abc/[string path]/[int path1]/[string path2]/[int ...path3] () returns int = @java:Method {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/variable_number_of_resource_paths.bal:36]: Better to add tests having usual function parameters too

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 7]
```diff
@@ -606,8 +610,41 @@ private void genJICallTerm(JIMethodCall callIns, int localVarOffset) {
...
             mv.visitFieldInsn(GETSTATIC, this.moduleInitClass, CURRENT_MODULE_VAR_NAME, GET_MODULE);
-            mv.visitMethodInsn(INVOKESPECIAL, BAL_ENV, JVM_INIT_METHOD,
-                               INIT_BAL_ENV, false);
+            mv.visitMethodInsn(INVOKESPECIAL, BAL_ENV, JVM_INIT_METHOD, INIT_BAL_ENV, false);
+        }
+
+        if (callIns.resourcePathArgs != null && !callIns.resourcePathArgs.isEmpty()) {
+            List<BIROperand> pathArgs = callIns.resourcePathArgs;
+            int pathVarArrayIndex = this.indexMap.addIfNotExists("$pathVarArray", symbolTable.anyType);
+            int bundledArrayIndex = this.indexMap.addIfNotExists("$bundledPathArgs", symbolTable.anyType);
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:619]:
```suggestion
            int bundledArrayIndex = this.indexMap.addIfNotExists("$pathArrayArgs", symbolTable.anyType);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 8]
```diff
@@ -183,6 +187,55 @@ private boolean hasEqualParamCounts(JMethodRequest jMethodRequest, JMethod jMeth
...
+        return false;
+    }
+
+    private void bundlePathParams(JMethodRequest jMethodRequest) {
+        if (jMethodRequest.pathParamSymbols.isEmpty()) {
+            return;
+        }
+        List<BType> paramTypes = new ArrayList<>(Arrays.asList(jMethodRequest.bParamTypes));
+        int initialPathParamIndex = paramTypes.indexOf(jMethodRequest.pathParamSymbols.get(0).type);
+        paramTypes.removeAll(jMethodRequest.pathParamSymbols.stream().map(BVarSymbol::getType)
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:232]: lets not use any streams

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 9]
```diff
@@ -183,6 +187,55 @@ private boolean hasEqualParamCounts(JMethodRequest jMethodRequest, JMethod jMeth
...
+                    .isAssignableFrom(paramTypes[1])) {
+                // This is for object interop functions when self is passed as a parameter
+                if (jMethod.isBalEnvAcceptingMethod()) {
+                    return true;
+                }
+                jMethod.setReceiverType(jMethodRequest.receiverType);
+                bundlePathParams(jMethodRequest);
+                jMethod.hasBundledPathParams = true;
+                return jMethodRequest.receiverType != null;
+            } else if (count == reducedParamCount + 2 && this.classLoader.loadClass(BArray.class.getCanonicalName())
```
Comment (Nadeeshan96) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:210]:
```suggestion
            } else if ((count == reducedParamCount + 2) && this.classLoader.loadClass(BArray.class.getCanonicalName())
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java [hunk 10]
```diff
@@ -742,4 +743,8 @@ static void bar() {
             throw ErrorCreator.createError(StringUtils.fromString("error!!!"));
         });
     }
+
+    public static int getResource(BArray paths) {
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java:747]: Lets add another test with Environment as well

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java (hunk 9 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:210]: Better to add tests covering these lines

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/ResourcePathArgumentsTest.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Test cases for java interop functions with variable number of resource path params.
+ *
+ * @since 2201.5.0
+ */
+
+public class ResourcePathArgumentsTest {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/ResourcePathArgumentsTest.java:33]:
```suggestion
public class ResourcePathArgumentsTest {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/variable_number_of_resource_paths.bal [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import ballerina/jballerina.java;
+import ballerina/test;
+
+public client isolated class Client {
+
+    public isolated function init() {
+
+    }
+
+    isolated resource function get abc/[string ...path] () returns int = @java:Method {
```
Comment (Nadeeshan96) [inline tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/variable_number_of_resource_paths.bal:26]:
```suggestion
    isolated resource function get abc/[string ...path]() returns int = @java:Method {
```

Comment (Nadeeshan96) [reply tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/variable_number_of_resource_paths.bal:26]: Remove the spaces in below occurrences too. See https://ballerina.io/learn/style-guide/top-level-definitions/#function-definition & https://ballerina.io/spec/http/#2423-resource-methods

**Comment by gabilang** [comment]
Closing this since the fix merged into master with https://github.com/ballerina-platform/ballerina-lang/pull/39898

**Pull request closed without merging**